### PR TITLE
Fix :  "attempt to index a number value"

### PR DESCRIPTION
**Pull request by ls-devs**
Fix "attempt to index a number value" on latest neovim build.

### CODE REVIEW

**Comment by dsully** [comment]
This fixes `nvim-notify` on nightly for me as well.

**Comment by GitMurf** [comment]
what is best way to apply this until the repo maintainer merges the PR? is the only real option to manually go replaces the file in my local installation? Thanks!

**Comment by ls-devs** [comment]
> what is best way to apply this until the repo maintainer merges the PR? is the only real option to manually go replaces the file in my local installation? Thanks!

You can just change `rcarriga/nvim-notify` by `ls-devs/nvim-notify` in your nvim config where you call the plugin I guess ?

**Comment by dpetka2001** [comment]
This PR is incomplete. It doesn't work for Neovim 0.9.5 (latest stable). If it gets merged like this, users who use 0.9.5 will have notifications broken. Notifications work fine with the current state of `lua/notify/stages/util.lua` file. I would suggest adding guards against the version of Neovim being used. Something like
```lua
if vim.fn.has("nvim-0.10") then
  local exists, cur_win_conf = util.get_win_config(win)
else
  local exists, cur_win_conf = pcall(vim.api.nvim_win_get_config, win)
end
```
The same should be done for the rest of the changes made in this PR.

Maybe there is a better way to do this, but I believe you get the idea.

**Comment by ls-devs** [comment]
@dpetka2001 Thank you for pointing it out.
For now I'll commit with guards everywhere it's needed. If anyone have an idea of how to do it more clean let us know.

**Comment by ls-devs** [comment]
@dpetka2001 Let me know if this commit fix your issue.
But, tbh, I don't know if this is the right way to do it. There must be a cleaner way to achieve this without using `if ... else` everywhere in the code.

**Comment by dpetka2001** [comment]
Did you test it? For me it doesn't solve the issue. I created a branch in my forked repo that appears to solve it for both nightly and stable. [Here](https://github.com/dpetka2001/nvim-notify/commit/d590b4d5b24039887ebb3fc9087306f73d0ad43e) you can see the commit with the differences. Please feel free to incorporate the changes.

**Comment by dpetka2001** [comment]
@xfzv You're using the master branch which is the initial state of the fork. Use the `fix/index_value` branch like so
```lua
  {
    "dpetka2001/nvim-notify",
    branch = "fix/index_value",
  },
```

**Comment by ls-devs** [comment]
> Did you test it? For me it doesn't solve the issue. I created a branch in my forked repo that appears to solve it for both nightly and stable. [Here](https://github.com/dpetka2001/nvim-notify/commit/d590b4d5b24039887ebb3fc9087306f73d0ad43e) you can see the commit with the differences. Please feel free to incorporate the changes. In your last commit you appear to make changes in the `window_intervals` function as well, but I believe that's not needed. Only the `slot_after_previous` function needs changes.

I tested it. But was in a hurry so I might no use the right `nvim` version. I rectified with your changes. Please tell me if it works because its working for me.

**Comment by latipun7** [comment]
The last change works for me too 👍

**Comment by dpetka2001** [comment]
@ls-devs Yes the latest changes solve the problem for both nightly and stable from the testing that I did. I would like to suggest that other people test it both on nightly and stable if possible, so that we haven't overlooked any edge case.

**Comment by dpetka2001** [comment]
@xfzv Please continue to use the branch of @ls-devs which is also associated with this PR. I will delete mine, since there's no purpose existing any more. I only created it to solve the problem for Neovim 0.9.5 which the initial state of the PR didn't take into account.

**Comment by iwansyahp** [comment]
in case there is someone who use LazyVim and encounter same issue, you can use proposed solution here by creating a file under `~/.config/nvim/lua/plugins`

`nvim-notify.lua`
```
return {
  {
    "ls-devs/nvim-notify",
    keys = {
      {
        "<leader>un",
        function()
          require("notify").dismiss({ silent = true, pending = true })
        end,
        desc = "Dismiss all Notifications",
      },
    },
    opts = {
      timeout = 3000,
      max_height = function()
        return math.floor(vim.o.lines * 0.75)
      end,
      max_width = function()
        return math.floor(vim.o.columns * 0.75)
      end,
      on_open = function(win)
        vim.api.nvim_win_set_config(win, { zindex = 100 })
      end,
    },
    init = function()
      -- when noice is not enabled, install notify on VeryLazy
      if not require("lazyvim.util").has("noice.nvim") then
        require("lazyvim.util").on_very_lazy(function()
          vim.notify = require("notify")
        end)
      end
    end,
  },
}
```

**Comment by dpetka2001** [comment]
@iwansyahp I believe you don't need all this. With just 
```lua
return {
  {
    "ls-devs/nvim-notify",
    branch = "fix/fix_index_value",
  },
}
```
You should be ok. The `opts` from the LazyVim part should still be merged with yours.

**Comment by dpetka2001** [comment]
@ls-devs I like @kawre's idea and seems more maintainable as well. It's up to you to make the final decision to incorporate it or not.

File: lua/notify/stages/util.lua [hunk 1]
```diff
@@ -142,12 +142,17 @@ end
 function M.slot_after_previous(win, open_windows, direction)
   local key = slot_key(direction)
   local cmp = is_increasing(direction) and less or greater
-  local exists, cur_win_conf = pcall(vim.api.nvim_win_get_config, win)
+  local exists, cur_win_conf
+  if vim.version().minor < 10 then
+    exists, cur_win_conf = pcall(vim.api.nvim_win_get_config, win)
+  else
+    exists, cur_win_conf = util.get_win_config(win)
+  end
```
Comment (dpetka2001) [inline lua/notify/stages/util.lua:150]:
```suggestion
local exists, cur_win_conf = pcall(vim.api.nvim_win_get_config, win)
```
I would also suggest to leave this as it was, since it doesn't affect the problem at hand.

**Review by dpetka2001 — CHANGES_REQUESTED** [review]

File: lua/notify/stages/util.lua (hunk 1 above)
Comment (ls-devs) [reply lua/notify/stages/util.lua:150]: Of course @kawre idea is good and I'll take time to implement it tomorrow.

> I would also suggest to leave this as it was, since it doesn't affect the problem at hand.

The only reason I did this is because there is already an util function that does the same thing, so I thought it was a good idea to reuse what already exists.

**Comment by mehalter** [comment]
Would it also be safer to do 

```lua
local function get_slot(conf, _key)
  return vim.fn.has "nvim-0.10" == 1 and conf[_key] or conf[_key][false]
end
```

That way it protects against whenever 1.0 comes out and the minor version goes below 10 again. Not that that's happening any time soon, but just for down the line.

**Comment by ls-devs** [comment]
@mehalter You are totally true, I didn't thought about this kind of issue. Will correct it tomorrow.
Thank you for pointing it out.

File: lua/notify/stages/util.lua (hunk 1 above)
Comment (dpetka2001) [reply lua/notify/stages/util.lua:150]: Yes, but the util function also uses a `pcall` to `vim.api.nvim_win_get_config` under the hood as well, so it's essentially the same thing. And both basically return the window handler at hand. Also there's no need to make a check here for the Neovim version. Only at the places where `key` is existent.

**Comment by rcarriga** [comment]
Thanks for the PR! :smile: I think the simplest solution here would be just switch both calls to `nvim_win_get_config` with `util.get_win_config` and then remove all `[false]` indexes (as if on >= 0.10.0). Am I missing something?

**Review by dpetka2001 — APPROVED** [review]

**Comment by dpetka2001** [comment]
Sorry for the wrong alert. I was under the assumption that when I make a suggestion I have to press `approve` for the PR to not be blocked. Still getting comfortable with Github and programming. Please disregard the `approved` notification by me.

**Comment by ls-devs** [comment]
> Thanks for the PR! 😄 I think the simplest solution here would be just switch both calls to `nvim_win_get_config` with `util.get_win_config` and then remove all `[false]` indexes (as if on >= 0.10.0). Am I missing something?

Yes that's was in my mind. I'll make the changes a fast as I can tomorrow so it can be reviewed and tested 👍🏼

**Comment by ls-devs** [comment]
@dpetka2001 , @rcarriga 

So I created `get_safe_slot`  function to handle all the case we need to remove `[false]` or not.
I also replaced `nvim_win_get_config` with `util.get_win_config`.
Would you please test it and tell me if it works for you ?

**Comment by wookayin** [comment]
I don't think we need such an utility --- needing to use such a field accessor on every single callsite would be unnecessarily complicated and error-prone. We can instead make the `win_conf` table (returned by `util.get_win_config`) always contain a number "float" for all the relevant fields, no matter which nvim version is used. In this way one can access the number always with `win_conf[key]`.

**Comment by ls-devs** [comment]
@wookayin Your idea seems to be the better solution for this issue. 
Would you please provide an example of how you would implement this ? I think I don't have enough knowledge about neovim to implement this alone.

**Comment by dpetka2001** [comment]
@ls-devs Based on what @rcarriga said, `util.get_win_config` seems to inherently manage this thing, so I don't believe you need the extra `slot` thing you added. I created an example branch based on what @rcarriga proposed. Take a look [here](https://github.com/dpetka2001/nvim-notify/commit/41e8dafa8eb39bbd4348cba2189c55394c780bb6).

**Comment by ls-devs** [comment]
> @ls-devs Based on what @rcarriga said, `util.get_win_config` seems to inherently manage this thing, so I don't believe you need the extra `slot` thing you added. I created an example branch based on what @rcarriga proposed. Take a look [here](https://github.com/dpetka2001/nvim-notify/commit/41e8dafa8eb39bbd4348cba2189c55394c780bb6). I tested it on both stable and nightly and it seems to work on my end.

Sorry, I misunderstood what he said ! I thought both was necessary.
I'll make a quickfix and push it.

**Comment by dpetka2001** [comment]
Please also test it on your end after your local changes, just to avoid something that I might have missed.

**Comment by ls-devs** [comment]
@dpetka2001 Can confirm it's working on both nightly and stable.
Sorry for the useless PRs and my misunderstanding of what @rcarriga said.

**Comment by dpetka2001** [comment]
No worries, I had also suggested in a previous comment to use the default Neovim API call instead of `util.get_win_config`, because I hadn't looked carefully into what it does until rcarriga proposed it. Then I looked again and realised that all these version checks were unnecessary, since `util.get_win_config` already makes that transformation for us and returns the window handler just like it's being returned from the Neovim API call in nightly as well.

**Comment by iwansyahp** [comment]
thanks @dpetka2001 for your answer here, I am new to https://github.com/rcarriga/nvim-notify/pull/253#issuecomment-1944364850

**Comment by rcarriga** [comment]
Awesome thanks for all the help everyone :grin:
